[Marcus Tang] iP

MarcusTXK opened the pull request:
# Biscuit

**Biscuit** is a good boi who will help you keep track of what you need to do.

Here a review from one of our satisfied customers:
> "Wow! Using Biscuit to manage tasks has revolutionised my life! I will never have to remember anything ever again!
> \- A totally legit CS2103T student

**Biscuit** is:
* Text-based
* Easy to learn
* ~~FAST~~ SUPER FAST to use

And best of all, it is **FREE**!

So what are you waiting for, download **Biscuit** now and you too can change your life in 4 simple steps:
1. Download it from [here](https://github.com/MarcusTXK/ip/releases/tag/A-Jar)
2. Double-click it
3. Add your tasks
4. Let it manage your LIFE for you 🐶

Features:
- [X] Managing different types of tasks
- [X] Quick search features to find tasks by names
- [ ] Search tasks by dates
---
When **Biscuit** is run, you should see the following output:
```
Woof from
████████████████████████████████████████
█▄─▄─▀█▄─▄█─▄▄▄▄█─▄▄▄─█▄─██─▄█▄─▄█─▄─▄─█
██─▄─▀██─██▄▄▄▄─█─███▀██─██─███─████─███
▀▄▄▄▄▀▀▄▄▄▀▄▄▄▄▄▀▄▄▄▄▄▀▀▄▄▄▄▀▀▄▄▄▀▀▄▄▄▀▀

      ████████████████████
    ██░░░░░░░░░░░░░░░░░░░░██
  ██░░░░██░░░░░░░░░░░░██░░░░██
  ██░░░░██░░██░░░░██░░██░░░░██
  ██░░████░░░░░░░░░░░░████░░██
    ██  ██░░░░████░░░░██  ██
        ██░░░░░██░░░░░██                  ██
          ██░░░░░░░░██░░████████        ██░░██
        ██░░████████░░░░░░░░░░░░██      ██░░██
      ██░░░░░░░░░░░░░░░░░░░░░░░░░░██  ██░░░░██
      ██░░░░░░░░░░░░░░██░░░░░░░░░░░░██░░░░██
      ██░░░░░░░░░░░░░░██░░░░░░░░░░░░██░░░░██
      ██░░░░░░██░░░░░░██░░░░░░░░░░░░░░░░██
      ██░░░░░░██░░░░░░██░░░░░░░░░░░░░░██
      ██░░░░░░██░░░░░░██░░░░░░░░░░████
████████████████████████████████████████

Woof! I'm Biscuit.
What can I do for you?
────────────────────────────────────────
```

Commit:
Add Gradle support

MarcusTXK committed:
Add skeletal version of Duke that greets and echos commands

MarcusTXK committed:
Add ability to store and display list to Duke

MarcusTXK committed:
Add Task class and Add ability to mark Task as done

MarcusTXK committed:
Add support for todos, deadline and events

MarcusTXK committed:
Add automated text testing, Update Duke name to Biscuit and Add javadocs

MarcusTXK committed:
Add error handling for all cases

MarcusTXK committed:
Add support for deleting tasks from the list and Update tests

MarcusTXK committed:
Update Biscuit to save tasks to csv file

MarcusTXK committed:
Update Deadline to use LocalDate and Update Event to use LocalDateTime

MarcusTXK committed:
Merge branch 'branch-Level-7'

MarcusTXK committed:
Resolve merge conflict between branch-Level-7 and branch-Level-8

MarcusTXK committed:
Refactor Biscuit code to be more OOP

MarcusTXK committed:
Organize classes into packages

MarcusTXK committed:
Add JUnit tests for ui and parser

MarcusTXK committed:
Add JavaDoc comments to all classes and methods

MarcusTXK committed:
Update code to comply with coding standard

MarcusTXK committed:
Add command for users to find tasks by keyword

MarcusTXK committed:
Merge branch 'branch-A-JavaDoc'

MarcusTXK committed:
Merge branch 'branch-A-CodingStandard'

MarcusTXK committed:
Merge branch 'branch-Level-9'

MarcusTXK committed:
Update code to follow Java coding standard

dillontkw commented on src/main/test/biscuit/parser/ParserTest.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Unit test for Parser class.
+ */
+class ParserTest {
```
Very detailed list of testing provided. Great job! 🙂

MarcusTXK replied:
Thank you!

dillontkw reviewed (commented):
Hi Marcus,

I loved how you personalised your project! It was easy to understand the flow and logic of your project. I didn't find any coding standard violations in your project, great job! 😄

MarcusTXK committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into add-gradle-support

MarcusTXK committed:
Update build.gradle for Biscuit and Update README.md description

MarcusTXK committed:
Merge branch 'add-gradle-support'

MarcusTXK committed:
Add checkStyle to detect coding violations and Fix detected violations

MarcusTXK committed:
Merge branch 'add-checkstyle-support'

MarcusTXK committed:
Add GUI to Biscuit using JavaFX

MarcusTXK committed:
Merge branch 'branch-level-10'

MarcusTXK committed:
Add ability for user to specify file path to save list through command line arguments

MarcusTXK committed:
Merge branch 'branch-A-Varargs'

Manoharan-Ajay-Anand commented on src/main/java/biscuit/commands/Command.java:
```diff
@@ -0,0 +1,65 @@
+        this.userInputs = userInputs;
+    }
+
+    /**
+     * Execute command from userInput.
+     *
+     * @param taskList Task list.
+     * @param ui       Ui to display.
+     * @param storage  Storage to save to.
+     * @return Response to user input.
+     * @throws BiscuitException Invalid input by user.
+     */
+    public abstract String execute(TaskList taskList, Ui ui, Storage storage) throws BiscuitException;
```
For the execute method, none of the Command sub-classes use the ui parameter and it is left unused. Generally, it is not a good practice to leave it dangling.

MarcusTXK replied:
Good catch! The ui parameter is a remnant of the text ui that I did not clean up yet.

Manoharan-Ajay-Anand commented on src/main/java/biscuit/parser/Parser.java:
```diff
@@ -0,0 +1,55 @@
+     * @return Command.
+     * @throws BiscuitException Invalid command by user.
+     */
+    public static Command parse(String userInput) throws BiscuitException {
+        String[] processedInputs = userInput.trim().split("\\s+", 2);
+        switch (processedInputs[0]) {
+        case "todo":
+            // Fallthrough
+        case "deadline":
+            // Fallthrough
+        case "event":
+            return new AddCommand(processedInputs);
+        case "done":
+            return new DoneCommand(processedInputs);
+        case "delete":
+            return new DeleteCommand(processedInputs);
+        case "list":
+            return new ListCommand(processedInputs);
+        case "find":
+            return new FindCommand(processedInputs);
+        case "bye":
+            return new ExitCommand(processedInputs);
+        default:
+            throw new BiscuitException("\u0ED2(\u25C9\u1D25\u25C9)\u096D OOPS!!! I'm sorry, but I don't know what that means...");
+        }
+    }
```
You did most of your parsing in the Command classes. The Command classes should only be concerned about managing the task list, storage and the UI. I would recommend parsing within your Parser class so that you can handle all input format related exceptions in a single place and adhere to the Separation of Concerns principle.

MarcusTXK replied:
Good point, will defintitely take this into consideration.

Manoharan-Ajay-Anand reviewed (commented):
Your code is quite neat and easy to follow. Cannot find any major violations. Remember to ensure that at the end of every file there is a newline. Intellij can do this automatically.

MarcusTXK commented:
Thanks @El0hime and @Manoharan-Ajay-Anand , appreciate the feedback! :)

MarcusTXK committed:
Add assertions to Biscuit

File path user passses in by commandline may be invalid.

This will cause Biscuit to be unable to save their list.

Lets do assertions to check the commandline input.

Assertions will stop the programme from running once it fails.

MarcusTXK committed:
Update build.gradle mainClassName config

MarcusTXK committed:
Update project to follow code quality guide

Some of the current code base does not follow code quality guidelines.

Let's update the code base to make it follow the guidelines.

This is so the code is easier to read and understand.

MarcusTXK committed:
Merge pull request #3 from MarcusTXK/branch-A-Assertions

Add assertions to Biscuit

MarcusTXK committed:
Merge branch 'master' into branch-A-CodeQuality

MarcusTXK committed:
Merge pull request #4 from MarcusTXK/branch-A-CodeQuality

Update project to follow code quality guide

MarcusTXK committed:
Merge branch 'master' into branch-A-Streams

MarcusTXK committed:
Add ability to mark done or delete multiple tasks at once

MarcusTXK committed:
Add github actions config to set up Continuous Integration

MarcusTXK committed:
Merge pull request #5 from MarcusTXK/branch-A-CI

Add github actions config to set up Continuous Integration

MarcusTXK committed:
Merge pull request #6 from MarcusTXK/branch-A-Streams

Add ability to mark done or delete multiple tasks at once

MarcusTXK committed:
Update gradle.yml file for github actions

MarcusTXK committed:
Update code to pass checkstyle

MarcusTXK committed:
Update Biscuit GUI

MarcusTXK committed:
Merge branch 'branch-A-BetterGui'

MarcusTXK committed:
Add screenshot of biscuit in docs folder

MarcusTXK committed:
Add a User Guide to docs\README.md

MarcusTXK committed:
Update README.md User Guide

MarcusTXK committed:
Set theme jekyll-theme-slate

MarcusTXK committed:
Update User Guide table and image url

MarcusTXK committed:
Update README.md and documentation

MarcusTXK committed:
Set theme jekyll-theme-cayman

MarcusTXK committed:
Update code documentation and Remove unused methods and classes
